### PR TITLE
chore(docs): Adds design docs for tactical

### PR DESCRIPTION
**Pull request by ttowncompiled**
Adds a design doc that outlines the concepts, advantages, and constraints of using the tactical library.

### CODE REVIEW

File: docs/tactical/DESIGN.md [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+
+Once supplied, Tactical will store the received JSON objects in persistent browser storage using JSON objects as keys. Keys are provided by the application when it makes calls to Tactical. Each key should be able to uniquely describe the object that’s associated with it. The representation of a key is up to the application, but can be as simple as a JSON representation of a standard HTTP request body. Tactical will handle translating each key to a unique token that can be used to index the key’s respective value.
+
+However, for most applications, this local representation will not be complete. By registering tactics, consumers will have the ability to easily specify what data would provide a more usable offline data model. From there, Tactical will handle prudently making the appropriate calls in order to eagerly prepare the offline model — reducing application latency in the process.
```
Comment (alxhub) [inline docs/tactical/DESIGN.md]: s/most/some/. I think that for many applications, this representation will work reasonably well, actually.

File: docs/tactical/DESIGN.md [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+### Data Manager
+
+Providing this experience requires that Tactical manage three separate representations of the application model: the complete data model _(which is managed by the backend)_, a partial data model _(which is managed by Tactical)_, and changes to the model _(which are managed by the end user)_. Mediating these three representations — in a way that both respects all expectations of the general model and achieves a desirable user experience — will be the most difficult challenge for this library.
+
+As a first step towards solving this challenge, Tactical provides a Data Manager. The Tactical Data Manager will be responsible for:
+
+1. Prudently loading data from the backend in order to build a better local model for the application to use.
```
Comment (alxhub) [inline docs/tactical/DESIGN.md]: This isn't actually a function of the data manager - prefetching/eager loading is entirely an application/tactic function.

File: docs/tactical/DESIGN.md [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+2. Reducing latency by supplying the application with cached data, and refreshing the cache when that data becomes too stale to be usable.
+3. Emitting data to the application when new versions are received from the backend or local store.
```
Comment (alxhub) [inline docs/tactical/DESIGN.md]: Add to this something about keeping track of which bits of data the application is displaying at any given point.

File: docs/tactical/DESIGN.md [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+Second, Firebase’s approach to data synchronization is that the last received request always wins. This means that if multiple users are trying to change the same object, then the last user to submit their change will overwrite all of the previous changes. Tactical takes the opposite approach. Using versioning, Tactical only accepts changes to the version that it has in store. Therefore, when multiple requests attempt to change the same version of an object, only the first request will be accepted. All further requests will be rejected and then Tactical will inform the application of the rejection so that it can resolve the request in a manner it deems the most appropriate.
```
Comment (alxhub) [inline docs/tactical/DESIGN.md]: Add to this paragraph (last sentence) something about how usually, that involves a simple merge to "rebase" the mutation on top of the latest version.

**Comment by alxhub** [comment]
LGTM pending the comments.

**Comment by ttowncompiled** [comment]
a69146a84039f080941eaf3ffdcefcbedd8669f3

**Pull request merged**